### PR TITLE
[Driver][SYCL] Turn on -fsycl-dead-args-optimization by default

### PR DESCRIPTION
**Pull request by mdtoguchi**
Turn on -fsycl-dead-args-optimization by default for -fsycl
compilations.  Update tests to reflect the new behavior

### CODE REVIEW

**Review by AGindinson — APPROVED** [review]
The change itself LGTM. Is there something in particular that has previously been blocking the "DAE on" default before, or is it just a leftover?

**Comment by vladimirlaz** [comment]
Failures in Jenkins/Precommit look like real regressions and they need to be fixed before submission of the PR.

File: clang/test/Driver/sycl-device-optimizations.cpp [hunk 1]
```diff
@@ -30,9 +30,9 @@
 // CHECK-NO-SYCL-EARLY-OPTS: "-fno-sycl-early-optimizations"
 
 /// Check that Dead Parameter Elimination Optimization is enabled
-// RUN:   %clang -### -fsycl -fsycl-dead-args-optimization %s 2>&1 \
+// RUN:   %clang -### -fsycl %s 2>&1 \
 // RUN:   | FileCheck -check-prefix=CHECK-DAE %s
-// RUN:   %clang_cl -### -fsycl -fsycl-dead-args-optimization %s 2>&1 \
+// RUN:   %clang_cl -### -fsycl %s 2>&1 \
```
Comment (AGindinson) [inline clang/test/Driver/sycl-device-optimizations.cpp:35]: Adding the CC1 call checks for the negative option may also be worth it, despite a few other tests rely on it.

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -695,6 +695,29 @@ static void addCoveragePrefixMapArg(const Driver &D, const ArgList &Args,
...
+
+    StringRef S(A->getValue());
+    unsigned OptLevel = 0;
+    if (S.getAsInteger(10, OptLevel))
+      return false;
+    return OptLevel > 1;
+  }
+  // No -O setting seen, default is -O2 for device.
+  return true;
+}
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/Clang.cpp:719]: Why is it that `-O4`/`-O0` require separate handling in the presence of a "general" block for `-O<n>` starting from the line 711? Is it something to do with different spelling?

File: clang/test/Driver/sycl-device-lib.cpp [hunk 3]
```diff
@@ -128,7 +128,7 @@
 
 /// ###########################################################################
 /// test llvm-link behavior for fno-sycl-device-lib
-// RUN: %clangxx -fsycl -fno-sycl-device-lib=all %s -### 2>&1 \
+// RUN: %clangxx -fsycl -fno-sycl-dead-args-optimization -fno-sycl-device-lib=all %s -### 2>&1 \
```
Comment (AGindinson) [inline clang/test/Driver/sycl-device-lib.cpp:131]: What exactly is the backstory for the no-device-libs mode?

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 4]
```diff
@@ -4738,9 +4761,10 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
     }
 
     // Turn on Dead Parameter Elimination Optimization with early optimizations
-    if (!(RawTriple.isAMDGCN()) &&
+    if (!(RawTriple.isAMDGCN()) && !Args.hasArg(options::OPT_fsycl_link_EQ) &&
         Args.hasFlag(options::OPT_fsycl_dead_args_optimization,
-                     options::OPT_fno_sycl_dead_args_optimization, false))
+                     options::OPT_fno_sycl_dead_args_optimization,
+                     isSYCLOptimizationO2orHigher(Args)))
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/Clang.cpp:4767]: Another thought on the `-O` level: instead of driver-level checks, could it be cleaner to avoid scheduling the pass within `clang/lib/CodeGen/BackendUtil.cpp` upon insufficient `OptimizationLevel::` value? Especially since the driver doesn't aim to diagnose the silent cancellation of DAE whenever it's less than `O2`.

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 5]
```diff
@@ -9133,8 +9157,10 @@ void SYCLPostLink::ConstructJob(Compilation &C, const JobAction &JA,
 
   // Turn on Dead Parameter Elimination Optimization with early optimizations
   if (!(getToolChain().getTriple().isAMDGCN()) &&
+      !TCArgs.hasArg(options::OPT_fsycl_link_EQ) &&
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/Clang.cpp:9160]: What is it that causes the incompatibility? Potential FPGA backend issues for archives have sprung to my mind, but _the truth_ is better than my vague hypothesis :)

**Review by AGindinson — COMMENTED** [review]
A couple suggestions & a few clarification points.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 2 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp:719]: They are separate from the other `-On` options (explicit in Options.td), so we need to special case them.

File: clang/test/Driver/sycl-device-lib.cpp (hunk 3 above)
Comment (mdtoguchi) [reply clang/test/Driver/sycl-device-lib.cpp:131]: The device libraries are pulled in by default now, which wasn't the case initially.  So instead of modifying the large number of tests that were created without the device libraries in mind, we added the `no-device-lib` to reduce test changes.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 4 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp:4767]: So would this be more of an exercise of the driver passing the enabling, and the BE would cancel the enabling if optimization level is less than -O2?  What about the case in which the user wants to force it to be turned on regardless of -O0 or not?  Is that valid?

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 5 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp:9160]: Thanks for calling this out - it may be the improper approach here.  I didn't fully realize the potential FPGA impact in doing the check this way.  We may just want to update the tests that are using `-fsycl-link` in separate compilation scenarios to disable DAE.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 4 above)
Comment (AGindinson) [reply clang/lib/Driver/ToolChains/Clang.cpp:4767]: > So would this be more of an exercise of the driver passing the enabling, and the BE would cancel the enabling if optimization level is less than -O2?

Yes, that's the essence of my suggestion.

> What about the case in which the user wants to force it to be turned on regardless of -O0 or not? Is that valid?

On whether such a scenario should be allowed, I don't know if there's a use case for that. In my understanding, the compiler users usually turn off the optimizations when they aim for better debuggability and compile-time performance. Can't say if our DAE implementation impacts the debuggability much - it probably doesn't - yet in both instances, users are generally prepared for the worsening of the runtime performance.

With that, if there's a particular use-case that requires such a "bypass" of the optimization level, it would truly be easier to implement this with the driver-level checks of O levels. Otherwise, we'd have to throw in some additional `CodeGenOpt` or somehow change the current one from `bool`-like to `enum`-like. But in any case, the current changes don't seem to allow such use-cases yet, and some CLI extension would probably be required (explicit passing of `-fsycl-dead-args-optimization` == enabling it always? additional flag? additional value `always` to `-fsycl-dead-args-optimization=`?).

To sum up, if there's a known use-case, let's enable it, and if there's none, my initial suggestion stands!

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp:4767]: We can look at this like any other optimization.  Some have abilities to disable explicitly, and some don't.  I think the ideal way to approach this is to have this optimization be turned on (without an implicit switch, based on optimization level in `CodeGenOpt`.  We would add a disabling option, which would be hooked to `-fno-sycl-dead-args-optimization` and have a 'forced' enabling if `-fsycl-dead-args-optimization` is passed on the command line.

Perhaps we can do that transform in the future, and keep this one as is, adding a TODO

Comment (bader) [reply clang/lib/Driver/ToolChains/Clang.cpp:4767]: > What about the case in which the user wants to force it to be turned on regardless of -O0 or not? Is that valid?

I don't think DAE will work with `-O0`. To make it work we need at least remove dead stores i.e. to enable DAE other optimizations must be executed first.

**Comment by mdtoguchi** [comment]
@AGindinson, I think your comments have been addressed, can you take another look?

**Review by AGindinson — APPROVED** [review]
LGTM with the current approach

**Comment by bader** [comment]
@intel/llvm-reviewers-runtime, ping.

**Comment by mdtoguchi** [comment]
@bader, the failures look to be performance related.  How do we proceed?

**Comment by bader** [comment]
> @bader, the failures look to be performance related. How do we proceed?

The failures do not seems to be blocking. I suggest we commit and investigate them separately.
